File: scripts/build-configs.mjs
import { readFileSync, readdirSync, mkdirSync, writeFileSync, existsSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const policyDir = join(root, "policies");
const generatedDir = join(root, "generated");

function readPolicies() {
  return readdirSync(policyDir)
    .filter((name) => name.endsWith(".md"))
    .sort()
    .map((name) => readFileSync(join(policyDir, name), "utf8").trim())
    .join("\n\n---\n\n");
}

function writeIfChanged(path, content) {
  if (existsSync(path) && readFileSync(path, "utf8") === content) return false;
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content, "utf8");
  return true;
}

const policies = readPolicies();
const template = readFileSync(join(root, "templates", "client-instructions.md"), "utf8");
const clients = {
  codex: "Codex",
  "claude-code": "Claude Code",
  chatgpt: "ChatGPT Manual Instructions",
};

for (const [slug, name] of Object.entries(clients)) {
  writeIfChanged(
    join(generatedDir, slug, "instructions.md"),
    template.replaceAll("{{CLIENT_NAME}}", name).replaceAll("{{POLICIES}}", policies),
  );
}

writeIfChanged(
  join(generatedDir, "codex", "config.toml"),
  "# Generated project-level Codex configuration.\n# Review before copying into .codex/config.toml.\nmodel_instructions_file = \"../generated/codex/instructions.md\"\n",
);
writeIfChanged(
  join(generatedDir, "codex", "AGENTS.md"),
  "# Agent Control Plane Global Guidance\n\nThis file is generated from the canonical policies in this repository.\nIt is intended for the managed block in ~/.codex/AGENTS.md.\n\n" + policies + "\n",
);
writeIfChanged(
  join(generatedDir, "claude-code", "README.md"),
  "# Claude Code adapter\n\nUse `generated/claude-code/instructions.md` as project memory content only after reviewing current Claude Code documentation.\nPrefer official `claude mcp` commands for MCP setup when available.\n",
);

File: scripts/build-configs.py
from __future__ import annotations

from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
POLICY_DIR = ROOT / "policies"
GENERATED_DIR = ROOT / "generated"


CLIENTS = {
    "codex": "Codex",
    "claude-code": "Claude Code",
    "chatgpt": "ChatGPT Manual Instructions",
}


def read_policies() -> str:
    parts: list[str] = []
    for path in sorted(POLICY_DIR.glob("*.md")):
        parts.append(path.read_text(encoding="utf-8").strip())
    return "\n\n---\n\n".join(parts)


def write_if_changed(path: Path, content: str) -> bool:
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists() and path.read_text(encoding="utf-8") == content:
        return False
    path.write_text(content, encoding="utf-8", newline="\n")
    return True


def main() -> int:
    policies = read_policies()
    template = (ROOT / "templates" / "client-instructions.md").read_text(encoding="utf-8")
    changed = False

    for slug, name in CLIENTS.items():
        content = template.replace("{{CLIENT_NAME}}", name).replace("{{POLICIES}}", policies)
        changed = write_if_changed(GENERATED_DIR / slug / "instructions.md", content) or changed

    codex_config = """# Generated project-level Codex configuration.
# Review before copying into .codex/config.toml.
model_instructions_file = "../generated/codex/instructions.md"
"""
    changed = write_if_changed(GENERATED_DIR / "codex" / "config.toml", codex_config) or changed

    global_agents = """# Agent Control Plane Global Guidance

This file is generated from the canonical policies in this repository.
It is intended for the managed block in ~/.codex/AGENTS.md.

""" + policies + "\n"
    changed = write_if_changed(GENERATED_DIR / "codex" / "AGENTS.md", global_agents) or changed

    claude_note = """# Claude Code adapter

Use `generated/claude-code/instructions.md` as project memory content only after reviewing current Claude Code documentation.
Prefer official `claude mcp` commands for MCP setup when available.
"""
    changed = write_if_changed(GENERATED_DIR / "claude-code" / "README.md", claude_note) or changed

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
